Introduce SiPhase2OuterTrackerLorentzAngle

mmusich opened the pull request:
#### PR description:

The primary goal of this PR is to introduce a new DB object to store per-`DetId` Lorentz Angle calibration values for the Phase-2 Outer Tracker. A new `CondFormat` called `SiPhase2OuterTrackerLorentzAngle` is introduced and used to populate the newly created `EventSetup` records `SiPhase2OuterTrackerLorentzAngleRcd` and `SiPhase2OuterTrackerLorentzAngleSimRcd`.
The former is then used to provide the calibration to the Outer Tracker local reconstruction via `Phase2StripCPEE`, while the latter is used to provide the simulation Lorentz Angle to the Phase2 Tracker digitizer concrete implementations for the OT (`PSPDigitizerAlgorithm`, `PSSDigitizerAlgorithm` and `SSDigitizerAlgorithm`).
   * The `EventSetup` records are for the moment populated via an ad-hoc `ESSource` called `SiPhase2OuterTrackerFakeLorentzAngleESSource` which can be used for both records, but eventually the input data will come from `CondDB` via `GlobalTag` once this PR is accepted. The `ESSource` is percolated via `fakePhase2OuterTrackerConditions_cff` to all the active phase-2 geometries.
   * The `ESSource` is configured in such a way to produce same values of the hardcoded Lorentz Angles used so far in the Simulation and Local Reconstruction (wich is &mu;<sub>H</sub>=0.07/T).     
   * Few utilities are added to fetch, import, dump, etc. the new data-base objects, as well as a DB object writer.
   * I profit of this PR to decouple the Phase-2 OT local reconstruction from the Phase-0 SiStrip dependent record `TkStripCPERecord` by introducing a new dependent record `TkPhase2OTCPERecord` which goes in the spirit of what was proposed originally at https://github.com/cms-sw/cmssw/pull/28101. The new dependent record at the moment depends only on the Phase Outer Tracker LA, but will depend on more records as more pieces are added to the local reconstruction. `TkPhase2OTCPERecord` is then subsequently percolated to the appropriate classes in Tracking.

The changes proposed here have been discussed at the Tracker Phase-2 simulation meeting of [May 22nd 2020](https://indico.cern.ch/event/879043/contributions/3874511/attachments/2043128/3422322/mm_phase2SimMeeting_22May2020.pdf), but have then been postponed to allow other urgent developments to enter the release.

#### PR validation:

The local branch passes the following workflows tests (one for each active geometry):

```
runTheMatrix.py --what upgrade -l 23206.0,23606.0,24406.0,24806.0,26606.0,27006.0,27406.0,27806.0,28206.0,28606.0,29006.0
```
as well as the unit tests of the respective packages involved.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This PR is not a backport and no backport is needed.
cc:
@emiglior @skinnari @tsusa @mtosi @JanFSchulte

mmusich committed:
Introduce SiPhase2OuterTrackerLorentzAngle and SiPhase2OuterTrackerLorentzAngleRcd,SiPhase2OuterTrackerLorentzAngleSimRcd

mmusich committed:
add serialization tests

mmusich committed:
improve SiPhase2OuterTrackerLorentzAngle payload writer

mmusich committed:
add amenities for fetching, importing, dumping, etc.

mmusich committed:
move Phase-2 Tracker records into their own files

mmusich committed:
modification to allow using SiPhase2OuterTrackerLorentzAngleSimRcd in Phase-2 Tracker Digitizer

mmusich committed:
modifications to allow using SiPhase2OuterTrackerLorentzAngleRcd into the Phase2 Tracker CPE estimation

mmusich committed:
code-checks & code-format

mmusich committed:
improve SiPhase2OuterTrackerLorentzAngleWriter

mmusich committed:
update autoCondPhase2 with LA for Outer Tracker

mmusich committed:
Introduce TkPhase2OTCPERecord and use it

mmusich committed:
clean-up CondFormat

mmusich committed:
clean-up CondFormat and change DB writer accordingly

mmusich committed:
move reading of OT conditions from sqlite file to prep-DB

mmusich committed:
add functions to retrieve LA for Pixel and Strip modules separately

mmusich committed:
clean-up

mmusich committed:
post-merge clean-up

mmusich committed:
introduce SiPhase2OuterTrackerFakeLorentzAngleESSourc

mmusich committed:
add a fake OT condition

mmusich committed:
improve fake OT LA conditions and use them everywhere

mmusich committed:
introduce TrackerGeometryUtils::getOuterTrackerDetIds and use it

mmusich committed:
simplify logic in Phase2StripCPE

mmusich committed:
fix steps as in #30988

mmusich committed:
fix mistake in CalibTracker/SiPhase2TrackerESProducers build file

mmusich committed:
add fakePhase2OuterTrackerConditions_cff also in new geometries D63,D64 and D65

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31095/17639

- This PR adds an extra 168KB to repository

- Found files with invalid states:
  - SLHCUpgradeSimulations/Geometry/python/fakeConditions_phase2OTCondition_cff.py:
    - Added: 32d0f3f7cf9452ff5ab9a16ea38846d3f3ab0e12
    - Deleted: 842f1135597b092d239388042a0bf6589611e3d5


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondCore/Utilities/plugins/Module_2XML.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondDBFetch.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondDBImport.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondFormats.h  modified in PR(s): #30886
  - File  Configuration/Geometry/python/dict2026Geometry.py  modified in PR(s): #31092, #30126

cmsbuild commented:
A new Pull Request was created by  @mmusich (Marco Musich) for master.

It involves the following packages:

CalibTracker/SiPhase2TrackerESProducers
CondCore/SiPhase2TrackerPlugins
CondCore/Utilities
CondFormats/DataRecord
CondFormats/SiPhase2TrackerObjects
CondTools/SiPhase2Tracker
Configuration/AlCa
Configuration/Geometry
Geometry/TrackerNumberingBuilder
RecoLocalTracker/Phase2TrackerRecHits
RecoLocalTracker/Records
RecoTracker/MeasurementDet
RecoTracker/Record
RecoTracker/TransientTrackingRecHit
SLHCUpgradeSimulations/Geometry
SimTracker/SiPhase2Digitizer
TrackingTools/Records


The following packages do not have a category, yet:

CalibTracker/SiPhase2TrackerESProducers
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@perrotta, @civanch, @Dr15Jones, @makortel, @cvuosalo, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @jpata, @tocheng, @slava77, @ggovi, @pohsun, @kpedro88 can you please review it and eventually sign? Thanks.
@echabert, @felicepantaleo, @pieterdavid, @robervalwalsh, @Martin-Grunewald, @alesaggio, @threus, @mmusich, @seemasharmafnal, @venturia, @vargasa, @makortel, @JanFSchulte, @dgulhan, @slomeo, @yduhm, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @tocheng, @mschrode, @ebrondol, @mtosi, @fabiocos, @gbenelli, @lecriste, @gpetruc this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 674894537d592d4fa25001241533b0675756597b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fb73b8/8655/summary.html
CMSSW: CMSSW_11_2_X_2020-08-07-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fb73b8/8655/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2612347
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cvuosalo commented on Geometry/TrackerNumberingBuilder/src/utils.cc:
```diff
@@ -15,3 +16,18 @@ std::vector<uint32_t> TrackerGeometryUtils::getSiStripDetIds(const GeometricDet&
+  for (const auto* dep : deepComp) {
+    const auto detId = dep->geographicalId();
+    if ((detId.det() == DetId::Detector::Tracker) && (detId.subdetId() > PixelSubdetector::PixelEndcap)) {
+      OTDetIds.push_back(detId);
```
```suggestion
      OTDetIds.emplace_back(detId);
```

cvuosalo replied:
`emplace_back` is preferred to `push_back`. I think there is even a Clang-tidy campaign that will make that conversion soon.

mmusich replied:
mmh... even for a simple native C++ type as `uint32_t`? I think the two variants are totally equivalent in such a case.

kpedro88 replied:
Indeed, I don't think it matters for primitive types.

cvuosalo replied:
It's OK in this case, so no change is necessary, but I think the main idea is there is almost never a reason to use `push_back`. `emplace_back` is either equivalent or superior, so it might as well always be used.

cvuosalo commented:
+1

civanch commented:
+1

kpedro88 commented on CondFormats/SiPhase2TrackerObjects/interface/SiPhase2OuterTrackerLorentzAngle.h:
```diff
@@ -0,0 +1,47 @@
+  SiPhase2OuterTrackerLorentzAngle(){};
+  ~SiPhase2OuterTrackerLorentzAngle(){};
+
+  inline void putLorentsAngles(std::unordered_map<unsigned int, float>& LA) { m_LA = LA; }
```
typo: Lorents -> Lorentz

kpedro88 commented on CondFormats/SiPhase2TrackerObjects/src/SiPhase2OuterTrackerLorentzAngle.cc:
```diff
@@ -0,0 +1,63 @@
+  std::unordered_map<unsigned int, float>::const_iterator id = m_LA.find(detid);
+  if (id != m_LA.end()) {
+    edm::LogError("SiPhase2OuterTrackerLorentzAngle") << "SiPhase2OuterTrackerLorentzAngle for DetID " << detid
+                                                      << " is already stored. Skippig this put" << std::endl;
```
typo: Skippig -> Skipping

kpedro88 commented on CondFormats/SiPhase2TrackerObjects/src/SiPhase2OuterTrackerLorentzAngle.cc:
```diff
@@ -0,0 +1,63 @@
+  return 0;
+}
+
+void SiPhase2OuterTrackerLorentzAngle::getLorentzAngles_PSP(const TrackerGeometry* trackerGeometry,
```
This function could be unified with the next two functions by adding `TrackerGeometry::ModuleType` as an extra argument

kpedro88 commented on CondFormats/SiPhase2TrackerObjects/src/SiPhase2OuterTrackerLorentzAngle.cc:
```diff
@@ -0,0 +1,63 @@
+  if (id != m_LA.end())
+    return id->second;
+  else {
+    edm::LogError("SiPhase2OuterTrackerLorentzAngle")
```
should this be a LogError or an exception?

kpedro88 commented on CondFormats/SiPhase2TrackerObjects/src/SiPhase2OuterTrackerLorentzAngle.cc:
```diff
@@ -0,0 +1,63 @@
+}
+
+void SiPhase2OuterTrackerLorentzAngle::printDebug(std::stringstream& ss, const TrackerTopology* /*trackerTopo*/) const {
+  std::unordered_map<unsigned int, float> detid_la = getLorentzAngles();
```
capture as const ref to avoid copy

kpedro88 commented on CondFormats/SiPhase2TrackerObjects/src/SiPhase2OuterTrackerLorentzAngle.cc:
```diff
@@ -0,0 +1,63 @@
+  size_t count = 0;
+  ss << "SiPhase2OuterTrackerLorentzAngleReader:" << std::endl;
+  ss << "detid \t Lorentz angle" << std::endl;
+  for (it = detid_la.begin(); it != detid_la.end(); ++it) {
```
use range-based loop

kpedro88 commented on CondTools/SiPhase2Tracker/plugins/SiPhase2OuterTrackerLorentzAngleWriter.cc:
```diff
@@ -0,0 +1,158 @@
+      << " There are " << detsLAtoDB.size() << " OT Lorentz Angle values assigned" << std::endl;
+
+  // SiStripLorentzAngle object
+  lorentzAngle = new SiPhase2OuterTrackerLorentzAngle();
```
This is a potential memory leak: `new ` is called in `analyze()`, but `delete ` is only called in the destructor
there doesn't appear to be any reason for `lorentzAngle` to be a class member at all (it's also not correctly indicated as a class member with leading `m_`)
could it just be allocated statically? if not, prefer to use smart pointers

kpedro88 commented on CondTools/SiPhase2Tracker/test/SiPhase2OuterTrackerLorentzAngleWriter_cfg.py:
```diff
@@ -0,0 +1,70 @@
+#! /usr/bin/env cmsRun
+# Author: Marco Musich (May 2020)
+from __future__ import print_function
+#import os
```
delete commented-out code

kpedro88 commented on RecoLocalTracker/Phase2TrackerRecHits/interface/Phase2StripCPE.h:
```diff
@@ -34,6 +38,8 @@ class Phase2StripCPE final : public ClusterParameterEstimator<Phase2TrackerClust
 
   const MagneticField& magfield_;
   const TrackerGeometry& geom_;
+  const SiPhase2OuterTrackerLorentzAngle& LorentzAngleMap_;
```
member variable names should start with lower case

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.h:
```diff
@@ -83,9 +84,12 @@ class Phase2TrackerDigitizerAlgorithm {
   edm::ESHandle<SiPixelLorentzAngle> SiPixelLorentzAngle_;
 
+  // Accessing Outer Tracker Lorentz angle from DB:
+  edm::ESHandle<SiPhase2OuterTrackerLorentzAngle> SiPhase2OTLorentzAngle_;
```
member variables should start with lower case (does not seem to be followed in this class...)

mmusich replied:
@suchandradutta please take note of this comment, I resolved a couple here.

kpedro88 commented:
@mmusich in addition to the (mostly minor) review comments, a few other things:
1. https://github.com/cms-sw/cmssw/blob/674894537d592d4fa25001241533b0675756597b/CalibTracker/SiPhase2TrackerESProducers/python/SiPhase2OuterTrackerFakeLorentzAngleESSource_cfi.py is an empty file. Is it needed? If not, it should be removed.
2. What is the purpose of this file? https://github.com/cms-sw/cmssw/blob/674894537d592d4fa25001241533b0675756597b/CondFormats/DataRecord/interface/SiPhase2OuterTrackerLorentzAngleRcd.h

mmusich committed:
Updated SiPhase2OuterTrackerLorentzAngle:
- correct typo
- throw when necessary
- use range-based loop
- remove redundant methods to retrieve LA for a certain module type

mmusich committed:
updated SiPhase2OuterTrackerLorentzAngleWriter:
- avoid memory leak
- remove unnecessary lines in the configuration file
- update description

mmusich committed:
update SiPhase2OuterTrackerFakeLorentzAngleESSource:
- correct typo in source code
- make the cfi meaningful

mmusich committed:
update Phase2StripCPE: use lower case for data members

mmusich committed:
update Phase-2 Tracker digitizer package: use lower-case for data members

cmsbuild commented:
The code-checks are being triggered in jenkins.

mmusich commented:
@kpedro88 thanks for the review.
I addressed all your comments a part from:

> What is the purpose of this file? https://github.com/cms-sw/cmssw/blob/674894537d592d4fa25001241533b0675756597b/CondFormats/DataRecord/interface/SiPhase2OuterTrackerLorentzAngled.h

This is essentially an aesthetic trick (taken from the phase-0 Strip setup, see [here](https://github.com/cms-sw/cmssw/search?q=SiStripCondDataRecords&unscoped_q=SiStripCondDataRecords) ) in order to have all the implementations of the records in a single file (`SiPhase2OuterTrackerCondDataRecords.h`), but still when including in some plugin, use an include instruction with a name suggestive of the actual record needed. I don't feel strongly about it, if you see any performance penalty I'll remove it.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31095/17681

- This PR adds an extra 216KB to repository

- Found files with invalid states:
  - SLHCUpgradeSimulations/Geometry/python/fakeConditions_phase2OTCondition_cff.py:
    - Added: 32d0f3f7cf9452ff5ab9a16ea38846d3f3ab0e12
    - Deleted: 842f1135597b092d239388042a0bf6589611e3d5


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondCore/Utilities/plugins/Module_2XML.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondDBFetch.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondDBImport.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondFormats.h  modified in PR(s): #30886
  - File  Configuration/Geometry/python/dict2026Geometry.py  modified in PR(s): #30126

cmsbuild commented:
Pull request #31095 was updated. @perrotta, @civanch, @Dr15Jones, @jpata, @jordan-martins, @chayanit, @cvuosalo, @wajidalikhan, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @makortel, @tocheng, @slava77, @ggovi, @pohsun, @kpedro88 can you please check and sign again.

kpedro88 commented on CondFormats/SiPhase2TrackerObjects/interface/SiPhase2OuterTrackerLorentzAngle.h:
```diff
@@ -0,0 +1,51 @@
+                                    const TrackerGeometry::ModuleType& theType,
+                                    std::unordered_map<unsigned int, float>& out) const;
+
+  void getLorentzAngles_PSP(const TrackerGeometry*, std::unordered_map<unsigned int, float>&) const;
+  void getLorentzAngles_PSS(const TrackerGeometry*, std::unordered_map<unsigned int, float>&) const;
+  void getLorentzAngles_2S(const TrackerGeometry*, std::unordered_map<unsigned int, float>&) const;
```
It looks like these are no longer defined anywhere, so could just be removed. (Or they could be implemented to refer to the unified `getLorentzAnglesByModuleType()`, if that's desirable.)

mmusich replied:
whoops, fixed now.

mmusich committed:
implement the getLorentzAngles_PSP, getLorentzAngles_PSS and getLorentzAngles_2S methods

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31095/17683

- This PR adds an extra 60KB to repository

- Found files with invalid states:
  - SLHCUpgradeSimulations/Geometry/python/fakeConditions_phase2OTCondition_cff.py:
    - Added: 32d0f3f7cf9452ff5ab9a16ea38846d3f3ab0e12
    - Deleted: 842f1135597b092d239388042a0bf6589611e3d5


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondCore/Utilities/plugins/Module_2XML.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondDBFetch.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondDBImport.cc  modified in PR(s): #30886
  - File  CondCore/Utilities/src/CondFormats.h  modified in PR(s): #30886
  - File  Configuration/Geometry/python/dict2026Geometry.py  modified in PR(s): #30126

cmsbuild commented:
Pull request #31095 was updated. @perrotta, @civanch, @Dr15Jones, @jpata, @jordan-martins, @chayanit, @cvuosalo, @wajidalikhan, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @makortel, @tocheng, @slava77, @ggovi, @pohsun, @kpedro88 can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: c90a4986aaec07614e6b4c8b81ae08c4452829c1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fb73b8/8694/summary.html
CMSSW: CMSSW_11_2_X_2020-08-10-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fb73b8/8694/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2612346
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cvuosalo commented:
+1

chayanit commented:
+1

perrotta commented:
+1
- Modifications to the reco code are consequential to the additions of the new DB objects, and also include the decoupling of the Phase-2 OT local reconstruction from the Phase-0 SiStrip dependent record: everything as described in the PR descriptions
- New parameters injected are identical to the previous default ones: no changes expected in outputs, none seen in the results of the jenkins tests

tlampen commented:
+1

kpedro88 commented:
+upgrade

civanch commented:
+1

pohsun commented:
+1

perrotta commented:
@cms-sw/db-l2 could you please review and sign, if you think so, before there will be any need to rebase this PR, which touches so many files?

ggovi commented:
+1

mmusich commented:
@smuzaffar, @silviodonato, @qliphy is the bot stuck? This PR is fully signed, but it does not show...

qliphy commented:
@mmusich   Not sure what happened to bot.  @ggovi  Would you mind to sign here again? Thanks!    
Anyway we can move this PR forward.

qliphy commented:
+1
